fixes Indentation on piping long string #200

jbodah opened the pull request:
* fixes #200

kassio commented:
Instead of changing the pattern, what do you think to use the `s:is_indentable_match()` to verify if the `=` isn't a string?

jbodah commented:
> Instead of changing the pattern, what do you think to use the s:is_indentable_match() to verify if the = isn't a string?

That sounds good. I'm not up to speed with all of the indent code yet. I'll give that a shot and post back with how it works

kassio commented:
any updates on this?

jbodah commented:
@kassio played with your idea a bit:

```vim
function! elixir#indent#indent_pipeline_assignment(ind, line)
  if a:line.current.text =~ s:STARTS_WITH_PIPELINE
        \ && elixir#util#is_indentable_match(a:line.last, s:LAST_NON_WHITESPACE_CHAR)
        \ && a:line.last.text =~ '^[^=]\+=.\+$'
    let b:old_ind.pipeline = indent(a:line.last.num)
    " if line starts with pipeline
    " and last line is an attribution
    " indents pipeline in same level as attribution
    return match(a:line.last.text, '=\s*\zs[^ ]')
  else
    return a:ind
  end
endfunction
```

But in testing I got issues like below

```
        def test do
          [ h | t ] = "a,b,c,d"
       -              |> String.split(",")
       -              |> Enum.reverse
       +  |> String.split(",")
       +  |> Enum.reverse

       -  { :ok, h }
       +{ :ok, h }
        end
```

jbodah commented:
Actually, think I see what's wrong...

jbodah commented:
@kassio how would you feel about doing something like:

1. find the last `=` in the previous line
2. check the syntax there
2a. if match operator, then return index of operator
2b. else, look at the next `=` in previous line
3. if none found, return original indent

kassio commented:
I was testing it locally, and I think that maybe this can solve the problem:
```viml
function! elixir#indent#indent_pipeline_assignment(ind, line)
  if a:line.current.text =~ s:STARTS_WITH_PIPELINE
        \ && a:line.last.text =~ '^[^=]\+=.\+$'
    let b:old_ind.pipeline = indent(a:line.last.num)
    " if line starts with pipeline
    " and last line is an attribution
    " indents pipeline in same level as attribution
    let assign_pos = match(a:line.last.text, '=\s*\zs[^ ]')
    return (elixir#util#is_indentable_at(a:line.last.num, assign_pos) ? assign_pos : a:ind)
  else
    return a:ind
  end
endfunction
```

kassio commented:
wdyt?

jbodah commented:
That looks better than what I had, but it still is open to weird edge cases (e.g. multiple `=`) and map arrows as below:

```ex
  it 'correctly indents pipelines on a map with "=>" syntax' do
    expect(<<~EOF).to be_elixir_indentation
    def build_command(input, output) do
      %{:hello => :world}
      |>
    end
    EOF
  end
```

What about changing the regex there to ignore arrows too?

kassio commented:
may I assume that it is an assign if, and only if, the `=` sign has anything different of `<,>,=,+,-` before/after and/or it's not a string?

jbodah commented:
Looking [here](http://elixir-lang.org/docs/master/elixir/operators.html) all of the following have `=`:

```ex
<=
>=
==
!= 
=~ 
=== 
!==
=
=>
```

The only other cases I can think of we'd want to ignore are strings and comments (I believe both are covered now)

kassio commented:
@jbodah agree with you, can you update the regex with these values then?

jbodah commented:
@kassio let me know how this looks

kassio commented on autoload/elixir/indent.vim:
```diff
@@ -22,6 +22,7 @@ let s:PAIR_START = '\<\%('.s:NO_COLON_BEFORE.s:BLOCK_START.'\)\>'.s:NO_COLON_AFT
 let s:PAIR_MIDDLE = '^\s*\%('.s:BLOCK_MIDDLE.'\)\>'.s:NO_COLON_AFTER.'\zs'
 let s:PAIR_END = '\<\%('.s:NO_COLON_BEFORE.s:BLOCK_END.'\)\>\zs'
 let s:LINE_COMMENT = '^\s*#'
+let s:MATCH_OPERATOR = '[^!><=]=[^~=>]'
```
Isn't this the `assign_operator`?

jbodah replied:
Technically it's the match operator I believe, but I'm up for calling it anything - let me know which you'd prefer http://elixir-lang.org/getting-started/pattern-matching.html#the-match-operator

kassio replied:
Oh, my bad. Totally forgot that this is elixir. 🏃
